torchtune/datasets/_preference.py

```diff
@@ -137,3 +140,136 @@ def _prepare_sample(self, sample: Mapping[str, Any]) -> Dict[str, List[int]]:
+      contributes to the loss.
+    - If ``train_on_input`` is False, the prompt is masked out (tokens replaced with -100).
+
+    Args:
+        tokenizer (ModelTokenizer): Tokenizer used by the model that implements the ``tokenize_messages`` method.
+        source (str): path to dataset repository on Hugging Face. For local datasets,
+            define source as the data file type (e.g. "json", "csv", "text"), pass
+            in the filepath in ``data_files``, and set ``split="train"``. See `Hugging Face's
+            <https://huggingface.co/docs/datasets/en/package_reference/loading_methods#datasets.load_dataset.path>`_
+            ``load_dataset`` for more details.
+        column_map (Optional[Dict[str, str]]): a mapping from the expected columns "chosen" and "rejected"
+            in the message transform :class:`~torchtune.data.ChosenRejectedToMessages` to the new column names in
+            the dataset. Keys should be "chosen" and "rejected" and values should be the actual column names.
+            If None, keep the default columns "chosen" and "rejected".
+        train_on_input (bool): Whether the model is trained on the prompt or not. Default is False.
+        new_system_prompt (Optional[str]): if specified, prepend a system message to every sample for both chosen
+            and rejected. This can serve as instructions to guide the model response. Setting this will OVERRIDE
+            any system messages already present in the dataset. Default is None.
+        split (str): ``split`` argument for ``datasets.load_dataset``. You can use this argument to load a subset
+            of a given split, e.g. ``split="train[:10%]"``. Default is "train".
+        **load_dataset_kwargs (Dict[str, Any]): additional keyword arguments to pass to ``load_dataset``.
+
+    Examples:
+
+    ::
+        my_preference_dataset.json
+        [
+            {
+                "chosen": [
+                    {
+                        "content": "What do I do when I have a hole in my trousers?",
+                        "role": "user"
+                    },
+                    { "content": "Fix the hole.", "role": "assistant" }
+                ],
+                "rejected": [
+                    {
+                        "content": "What do I do when I have a hole in my trousers?",
+                        "role": "user"
+                    },
+                    { "content": "Take them off.", "role": "assistant" }
+                ]
+            }
+        ]
+
+    ::
+
+    >>> from torchtune.datasets import preference_dataset
+    >>> dataset = preference_dataset(
+    ...     tokenizer=tokenizer,
+    ...     source="json",
+    ...     data_files=my_preference_dataset.json,
+    ...     train_on_input=False,
+    ...     split="train",
+    >>> )
+    >>> tokenizer.decode(dataset[0]["chosen_input_ids"], skip_special_tokens=True)
+    What do I do when I have a hole in my trousers?Fix the hole.
```

Comment:
We should fix this trim whitespace issue at some point...

Reply:
actually renders for me as `'user\n\nWhat do I do when I have a hole in my trousers?assistant\n\nTake them off.'` with llama3 tokenizer and `skip_special_tokens=True`, and `'[INST] What do I do when I have a hole in my trousers?  [/INST] Take them off.'` with llama2, but I wrote it this way for consistency with the other examples.